\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in,marginparwidth=2in]{geometry}

% use Unicode characters - try changing the option if you run into troubles with special characters (e.g. umlauts)

\usepackage[utf8]{inputenc}


% clean citations
\usepackage{cite}

% hyperref makes references clicky. use \url{www.example.com} or \href{www.example.com}{description} to add a clicky url
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% improves typesetting in LaTeX
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% text layout - change as needed
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in
\textheight 8.75in

% use adjustwidth environment to exceed text width (see examples in text)
\usepackage{changepage}

% adjust caption style

% remove brackets from references
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% headrule, footrule and page numbers
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}

% use \textcolor{color}{text} for colored text (e.g. highlight to-do areas)
\usepackage{color}

% define custom colors (this one is for figure captions)
\definecolor{Gray}{gray}{.25}

% this is required to include graphics
\usepackage{graphicx}

% use if you want to put caption to the side of the figure - see example in text
\usepackage{sidecap}

% use for have text wrap around figures
\usepackage{wrapfig}
\usepackage[pscoord]{eso-pic}
\usepackage[fulladjust]{marginnote}
\reversemarginpar

% document begins here
\begin{document}
\vspace*{0.35in}

% title goes here:
\begin{flushleft}
{\Large
  \textbf\newline{pyranges: performant pythonic genomicranges}
  % and arithmetic run length-encoding
}
\newline
% authors go here:
\\
Endre Bakken Stovner\textsuperscript{1}\textsuperscript{+},
Pål Sætrom\textsuperscript{1, 2},
\\
\bf{1} Department of
  Computer Science, Norwegian University
  of Science and Technology, Trondheim, 7013, Norway
\\
\bf{2} Department of Clinical and Molecular Medicine, Norwegian
  University of Science and Technology, Trondheim, 7013, Norway
\\
\bigskip
* endrebak85@gmail.com

\end{flushleft}

\section*{Abstract}

\textbf{Summary:} PyRanges is a high-performance datastructure for
representing and manipulating genomic ranges and their associated data in
Python. Since it is directly compatible with Python's wealth of high-performance
data science libraries we expect it to become a cornerstone of genomic analysis
in Python.

\textbf{Availability and Implementation:} PyRanges is available open-source under
the MIT license at https://github.com/endrebak/pyranges

\textbf{Contact:} endrebak85@gmail.com

\section*{Introduction}

Comparing sets of intervals is a fundamental task in genomics, and a few basic
operations allow for answering complex questions. Several very popular command
line tools for this purpose exist, like bedtools
\cite{doi:10.1093/bioinformatics/btq033} and bedops
\cite{doi:10.1093/bioinformatics/bts277}. There is also a wrapper for bedtools
which allows easy use from Python \cite{doi:10.1093/bioinformatics/btr539}. In
R, the foundational GenomicRanges library \cite{10.1371/journal.pcbi.1003118} is
used for feature comparisons, and it is a cornerstone of genomics packages in
the R BioConductor project.

Python is the fourth largest programming language in the world and it is widely
used in bioinformatics, yet Python lacks a GenomicRange implementation.

The GenomicRange is a datastructure for efficiently representing and operating
on genomic intervals. Each row contains a position (chromosome, start, end and
optionally, strand) and an arbitrary number of metadata columns.

Operations on pairs of GenomicRanges are especially useful. Then one list of
ranges can be used to query another to find e.g. an overlapping or the nearest
feature. So if one GenomicRange represents alignments and another represents
genomic annotations, a nearest search would find the genomic feature closest to
each alignment.

As the data in a PyRanges object is stored



\section*{Library}


PyRanges uses the nested containment list, originally implemented in the PyGr
library \cite{doi:10.1093/bioinformatics/btl647} for range based queries.

\section*{Discussion}

\section*{Conclusion}


\bibliography{library}

%This defines the bibliographies style. Search online for a list of available styles.
\bibliographystyle{abbrv}

\end{document}
